### PR TITLE
Tiny cleanup for Navigator code (without introducing dependency)

### LINKED ISSUE #125724: Cleanup for Navigator code
The `navigator.dart` has ~10 repeats of things like:

```dart
    final _RouteEntry? currentRouteEntry = _navigator!._history.cast<_RouteEntry?>().lastWhere(
      (_RouteEntry? e) => e != null && _RouteEntry.isPresentPredicate(e),
      orElse: () => null,
    );
```

while it can be greatly simplified as:

```dart
    final _RouteEntry? currentRouteEntry = _navigator!._history.lastWhereOrNull(_RouteEntry.isPresentPredicate);
```

Thus, it seems that we can beautify the code a little bit.

### PR DESCRIPTION
**Pull request by fzyzcjy**
Close https://github.com/flutter/flutter/issues/125724

> The `navigator.dart` has ~10 repeats of things like:
> 
> ```dart
>     final _RouteEntry? currentRouteEntry = _navigator!._history.cast<_RouteEntry?>().lastWhere(
>       (_RouteEntry? e) => e != null && _RouteEntry.isPresentPredicate(e),
>       orElse: () => null,
>     );
> ```
> 
> while it can be greatly simplified as:
> 
> ```dart
>     final _RouteEntry? currentRouteEntry = _navigator!._history.lastWhereOrNull(_RouteEntry.isPresentPredicate);
> ```
> 
> Thus, it seems that we can beautify the code a little bit.

Same as https://github.com/flutter/flutter/pull/125099, but without external dependency

For more detailed explanation why it does not introduce external dependency: https://github.com/flutter/flutter/pull/125099#discussion_r1174230502

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by Piinks** [comment]
Hey @fzyzcjy is there an issue this is resolving? Can you update the description of the PR with an explanation of what this change is doing?

**Comment by fzyzcjy** [comment]
@Piinks Hi, it is solving the same issue as https://github.com/flutter/flutter/pull/125099 mentions. The former issue was closed because hixie thought that the external dependency should not be introduced, hence this PR.

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 1]
```diff
@@ -5739,3 +5710,24 @@ class RestorableRouteFuture<T> extends RestorableProperty<String?> {
 
   static NavigatorState _defaultNavigatorFinder(BuildContext context) => Navigator.of(context);
 }
+
+extension <T> on Iterable<T> {
+  T? firstWhereOrNull(bool Function(T element) test) {
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/navigator.dart:5715]: flutter style guide against extension, can you create a private method that takes a iterable and test function instead?

**Comment by Piinks** [comment]
> it is solving the same issue as https://github.com/flutter/flutter/pull/125099 mentions.

Can you update the PR description to link to the issue and explain the fix? Having to backtrack through other PRs and issues is difficult to follow. :)

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 2]
```diff
@@ -5711,23 +5711,25 @@ class RestorableRouteFuture<T> extends RestorableProperty<String?> {
...
 
-extension <T> on Iterable<T> {
-  T? firstWhereOrNull(bool Function(T element) test) {
-    for (final T element in this) {
-      if (test(element)) {
-        return element;
-      }
+/// First item satisfying the predicate, or null if not found.
+@visibleForTesting
+T? iterableFirstWhereOrNull<T>(Iterable<T> iterable, bool Function(T element) test) {
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/navigator.dart:5716]: make it private?

Comment (chunhtai) [inline packages/flutter/lib/src/widgets/navigator.dart:5716]: just hard code this to _RouteEntry ? also maybe rename this to
_firstRouteEntryWhereOrNull

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/navigator.dart:5716]: and typedef the test method function signature

**Comment by Hixie** [comment]
test-exempt: refactor (in case the test is removed)

**Comment by fzyzcjy** [comment]
@Piinks Done

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 2 above)
Comment (fzyzcjy) [reply packages/flutter/lib/src/widgets/navigator.dart:5716]: Then IMHO we cannot test it in navigator_test.dart?

Comment (fzyzcjy) [reply packages/flutter/lib/src/widgets/navigator.dart:5716]: > and typedef the test method function signature

May I know it a bit more?

```
class _RouteEntry {
  _firstRouteEntryWhereOrNull() {...}
}
```

Then I wonder how to test it using typedef

Comment (fzyzcjy) [reply packages/flutter/lib/src/widgets/navigator.dart:5716]: IMHO then I cannot unit-test this function inside navigator_test.dart

Comment (fzyzcjy) [reply packages/flutter/lib/src/widgets/navigator.dart:5716]: Oh I see, you mean the argument named `test`, not "a test to the newly added function"

**Comment by fzyzcjy** [comment]
@chunhtai done

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 3]
```diff
@@ -4045,12 +4036,10 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
...
     // Announce route name changes.
     if (widget.reportsRouteUpdateToEngine) {
-      final _RouteEntry? lastEntry = _history.cast<_RouteEntry?>().lastWhere(
-        (_RouteEntry? e) => e != null && _RouteEntry.isPresentPredicate(e), orElse: () => null,
-      );
+      final _RouteEntry? lastEntry = _lastRouteEntryWhereOrNull(_RouteEntry.isPresentPredicate);
       final String? routeName = lastEntry?.route.settings.name;
       if (routeName != null && routeName != _lastAnnouncedRouteName) {
-        SystemNavigator.routeInformationUpdated(uri: Uri.parse(routeName));
+        SystemNavigator.routeInformationUpdated(location: routeName);
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/navigator.dart:4042]: revert this? the location is deprecated

**Review by chunhtai — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 3 above)
Comment (fzyzcjy) [reply packages/flutter/lib/src/widgets/navigator.dart]: (not sure why this happened, maybe a merge error)

**Review by Piinks — APPROVED** [review]
LGTM thank you!

**Pull request merged**